### ISSUE TITLE
Mids stored in `Tunnel` objects are unreliable and incorrect

### ISSUE THREAD

**Issue by devos50** [Tribler/py-ipv8#166]
My recent PR on Tribler made the payouts based on the mid of a specific peer. However, it turns our that the `mid` property, stored in `Tunnel` objects (circuits, relays, and exit sockets) is not always correct. This in turn sometimes make the payout process fail over end-to-end circuits.

After some debugging, I found out that the `CreatePayload` sends a public key. This is also according to the anonymous downloading specifications as described [here](https://github.com/Tribler/tribler/wiki/Anonymous-Downloading-and-Streaming-specifications). In particular, when peer A initiates the creation of a 1-hop circuit, they send a `create` message to the exit node, which contains _the public key of the exit node itself_. However, when the exit node parses this create message, [the code treats this public key like it was the public key of the sender of the create message](https://github.com/Tribler/py-ipv8/blob/master/ipv8/messaging/anonymization/community.py#L842). The public key and socket address of the `Peer` object created in this line, do not match. This error amplifies since [the mid of a `TunnelExitSocket` object is used to build an e2e circuits](https://github.com/Tribler/py-ipv8/blob/master/ipv8/messaging/anonymization/hidden_services.py#L411). As a result, payouts for end-to-end circuits fail.

It is trivial to reproduce this issue by running the following unit test:

```
@twisted_wrapper
def test_create_circuit(self):
    self.nodes[1].overlay.settings.become_exitnode = True
    yield self.introduce_nodes()

    self.nodes[0].overlay.build_tunnels(1)

    yield self.deliver_messages()

    # Check the keys
    exit_socket = self.nodes[1].overlay.exit_sockets.values()[0]
    self.assertEqual(exit_socket.sock_addr, self.nodes[0].overlay.my_peer.address)
    self.assertEqual(exit_socket.mid, self.nodes[0].overlay.my_peer.mid)
```

So how can we solve this? In particular, I think that the exit node needs the public key of the previous node. Should we send this in a `create` message? If so, we break backward compatibility. Or since we know the socket address of the originator of a `create` message, maybe we can use the `get_peers` method to get the public key? Or is there some other solution?

@qstokkink @egbertbouman could you please comment on this issue?

**Comment by ichorid** [comment]
It seems to me that the specifications of the anonymous downloading protocol were correctly implying that peer A (the initiator) should send *its own* public key in the message. However, the specification writing was a little bit ambiguous, and whoever implemented the protocol interpreted this ambiguity in the wrong way. That's why this field is treated by the exit node like a public key of the sender.

Two questions arise now:
1. How was it able to work to this day?
2. What happens if we change the "create" message to contain *initiator's* public key (as it was apparently originally designed)?

In essence, there are 4 possibilities:
1. The specification was correct, implementation was not: we fix it to work according to the specification, and it works, and compatibility is still there.
2. The specification was incorrect, implementation was correct: we *have* to break backwards compatibility to fix it.
3. The specification was incorrect, implementation was incorrect: this is the toughest case. We obviously have to break compatibility.
4. The specification was correct, but there are *two mutually neutralizing bugs* in the implementation. Obviously, we have to break compatibility.

So, the obvious first step should be to just change the message to contain the sender's PK, cross fingers and hope for the best...

**Comment by qstokkink** [comment]
The fundamental problem is that a `cell` packet is unsigned: https://github.com/Tribler/py-ipv8/blob/master/ipv8/messaging/anonymization/community.py#L779 . This was also never required before payouts.

So if we want to know the relayer's key we will either need to start signing the cells or including it in the payload. Either way, as @ichorid points out, we will need to make a breaking change.

**Comment by qstokkink** [comment]
Resolved by #167
